### PR TITLE
Refactor schema generation

### PR DESCRIPTION
**Pull request by weinbe58**
Code Gen for various parts of the schema create a `Results` object that stores the main results for the codegen.

These objects have some internal API's that make it useful for composing results objects together, namely `slice` and `append` as these are the two things that need to be done when composing the pwl and pwc sequences together during code generation.

### CODE REVIEW

**Comment by weinbe58** [comment]
In the next PR I/John intend to tackle the error messages. Currently we do a validation and code generation in the same sweep but the next implementation will split the errors into a separate scan while keeping the code written here as just the emission of the task IR.

File: src/bloqade/codegen/hardware/piecewise_linear.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+
+def check_continiuity(left, right):
+    if left != right:
+        diff = abs(left - right)
+        raise ValueError(
+            f"discontinuity with a jump of {diff} found when compiling to "
+            "piecewise linear."
+        )
+
```
Comment (weinbe58) [inline src/bloqade/codegen/hardware/piecewise_linear.py:91]: Move inside codegen

File: src/bloqade/codegen/hardware/piecewise_linear.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        duration = ast.duration(**self.assignments)
+
+        if len(ast.coeffs) == 0:
+            start = Decimal(0)
+            stop = Decimal(0)
+        elif len(ast.coeffs) == 1:
+            start = ast.coeffs[0](**self.assignments)
+            stop = start
+        elif len(ast.coeffs) == 2:
+            duration = ast.duration(**self.assignments)
```
Comment (weinbe58) [inline src/bloqade/codegen/hardware/piecewise_linear.py:133]:
```suggestion
```

File: src/bloqade/codegen/hardware/quera.py [hunk 3]
```diff
@@ -1,364 +1,307 @@
...
+            if self.global_detuning is None:
+                self.global_detuning = PiecewiseLinear(
+                    [Decimal(0), duration], [Decimal(0), Decimal(0)]
+                )
+
+        if self.local_detuning is None:
+            pass
 
     @staticmethod
     def convert_position_to_SI_units(position: Tuple[Decimal]):
```
Comment (weinbe58) [inline src/bloqade/codegen/hardware/quera.py:307]: Move distance unit conversion to Results object.

File: src/bloqade/codegen/hardware/piecewise_constant.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                (f"end time for slice {stop_time} is larger than duration {duration}.")
+            )
+
+        if stop_time < start_time:
+            raise ValueError(
+                (
+                    f"end time for slice {stop_time} is smaller than "
+                    f"starting value for slice {start_time}."
+                )
+            )
```
[the remark this answers is not in the document]
Comment (weinbe58) [reply src/bloqade/codegen/hardware/piecewise_constant.py:180]: already exists

**Review by johnzl-777 — APPROVED** [review]

**Pull request merged**